### PR TITLE
[FLINK-25372] Add thread dump feature for jobmanager

### PR DESCRIPTION
**Pull request by X-czh**
## What is the purpose of the change

Add thread dump feature for jobmanager, so use can get jobmanager thread information easily, which is useful for debugging job submission and scheduling issues especially in OLAP scenarios.

jm_threaddump

## Brief change log

- Move ThreadDumpInfo out from taskmanager-specific package
- Add REST API for requesting JobManager thread-dump
- Add thread dump feature for JobManager on Web UI

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) yes
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) no need for documentation

### CODE REVIEW

**Comment by X-czh** [comment]
@flinkbot run azure

**Comment by X-czh** [comment]
@xintongsong Thanks for helping review this MR. I tried to run azure CI, but failed twice due to being stuck in the compiling phase (checksum validation failed for different packages between two runs though):

screenshot-20211220-165915

However, I was able to build it locally, how could I debug this problem? The CI log doesn't seem to be helpful to me.

**Comment by xintongsong** [comment]
@X-czh
No worries. There seems to be some infrastructure problems. The community is currently looking into this. We can re-run the CI after the problem being fixed.

**Comment by X-czh** [comment]
@flinkbot run azure

**Review by xintongsong — COMMENTED** [review]
@X-czh,
I've reviewed the runtime & rest changes, and they look quite good to me. There's only one minor comment. I also tried this feature locally and it works fine for me. I'll try to find someone with frontend expertise to double check on the codes.

File: flink-runtime-web/web-dashboard/src/app/pages/job-manager/thread-dump/job-manager-thread-dump.component.ts
Comment (yangjunhan) [inline flink-runtime-web/web-dashboard/src/app/pages/job-manager/thread-dump/job-manager-thread-dump.component.ts:67]: you need to set loading false in error as well. Another way is using `finalize` rxjs operator in the pipe so that you do not need to set loading false in two places.

**Review by yangjunhan — CHANGES_REQUESTED** [review]

**Comment by X-czh** [comment]
@xintongsong @yangjunhan Thanks for the fruitful comments. I've addressed all of them. Take a review when you are free ~

**Comment by xintongsong** [comment]
@X-czh,
I noticed that you have squashed all the changes into one commit. This makes the PR hard to review, and is not ideal for the maintaining of git history. Could you please change it back?
You can use `git commit --fixup <commit-to-be-fixed>` to address comments in separate commits. That helps reviewers identify new changes since previous review, and can be easily squashed later with `git rebase -i`.

**Comment by X-czh** [comment]
@xintongsong Thanks for the advice and sorry for the inconvenience. I'll try restoring the commits.

**Comment by X-czh** [comment]
@xintongsong I've restored the previous commits and addressed comments with fixup commits as you suggested. Take a look when you're free. Thx.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/ThreadDumpInfo.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/ThreadDumpInfo.java:56]: 1. It would be a good practice to stick to either constructors or factory methods for a class. Having both public constructor and factory methods could be confusing in long term.
2. When I suggested deduplicating with a static method in my earlier comment, I meant to deduplicate the logic "take a thread dump", which is not an exact replacement of a constructor which "create a ThreadDumpInfo with given thread information".

**Review by xintongsong — APPROVED** [review]
@X-czh,
Thanks for addressing the comment. I'm taking over from here. I've addressed my last comments myself, with a follow-up commit. Please take a look.

@yangjunhan,
Do you want to have another check?

If there's no other concerns, I'll merge this after AZP gives green light.

**Comment by X-czh** [comment]
@xintongsong The follow-up commit LGTM. It's my first time trying to contribute to the community, I did learn quite a few from this experience. Thanks for all you guys!

**Comment by xintongsong** [comment]
@flinkbot run azure

**Comment by xintongsong** [comment]
@X-czh 
As a first contribution, you did surprisingly well. Welcome to the community. 😄

**Review by yangjunhan — APPROVED** [review]
LGTM

**Comment by xintongsong** [comment]
The CI infrastructure has collapsed at the moment, and the major maintainers are absent due to Christmas. Let's wait a bit longer.

**Pull request closed without merging**
